File: python/CWE-22.py
import os

def read_file(filename):
    # Каталог, в котором должны находиться файлы
    base_directory = "/safe_directory/"
    
    # Путь к файлу, который необходимо прочитать
    file_path = os.path.join(base_directory, filename)
    
    try:
        with open(file_path, 'r') as file:
            return file.read()
    except FileNotFoundError:
        return "File not found."
    except Exception as e:
        return str(e)

# Входной файл запрашивается у пользователя
user_input = input("Enter the filename: ")

# Чтение содержимого файла
file_content = read_file(user_input)
print(file_content)

File: GO/CWE-22.go
package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"path/filepath"
)

func main() {
	http.HandleFunc("/readfile", readFileHandler)
	http.ListenAndServe(":8080", nil)
}

func readFileHandler(w http.ResponseWriter, r *http.Request) {
	filename := r.URL.Query().Get("file")
	if filename == "" {
		http.Error(w, "File parameter is missing", http.StatusBadRequest)
		return
	}

	// Опасное место: нет проверки пути
	filePath := filepath.Join("uploads", filename)

	data, err := ioutil.ReadFile(filePath)
	if err != nil {
		http.Error(w, "File not found", http.StatusNotFound)
		return
	}

	w.Write(data)
}

func init() {
	// Создание директории uploads и тестового файла
	os.Mkdir("uploads", 0755)
	ioutil.WriteFile("uploads/test.txt", []byte("This is a test file."), 0644)
	fmt.Println("Test file created at uploads/test.txt")
}
